### PR TITLE
feat: button with loader

### PR DESCRIPTION
**Pull request by pspaczek**
Add UiLoader to UiButton; 

Screens
Zrzut ekranu 2023-09-5 o 11 16 48
Zrzut ekranu 2023-09-5 o 11 04 39
Zrzut ekranu 2023-09-5 o 11 16 52
Zrzut ekranu 2023-09-5 o 11 17 10

### CODE REVIEW

File: src/components/atoms/UiButton/UiButton.vue
Comment (kubajmarek) [inline src/components/atoms/UiButton/UiButton.vue:1]: I think loading button should automatically also be `--is-disabled`.
It is stripping some control from the user, but I don't see a use case where we'd want loading button to still be active and actionable, and adding
```
:class="{ 'ui-button--is-disabled': isLoading }"
:isLoading
```
feels unnecessary.

**Comment by kubajmarek** [comment]
I'm wondering about the behaviour of disabled loading button - I'm not sure whether the loader should be greyed out. It's disabled, as a result of loading, so loading status is still "active".
Recent example of non-grayed button [here](https://www.figma.com/file/wz68QdkR6ldjmVrPYrazsr/Core-Triage-UI-Library?node-id=15584%3A126884&mode=dev).

File: src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue [hunk 1]
```diff
@@ -8,6 +8,7 @@
   <div
     ref="contentEl"
     :style="contentStyle"
+    style="display: contents"
```
Comment (kubajmarek) [inline src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue:11]: I don't think we should ever use non-dynamic `style` property. 👀

File: src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue [hunk 2]
```diff
@@ -123,7 +125,7 @@ const startTransition = () => {
   }
   onTransitionBeforeActive();
 };
-onMounted(() => {
+onMounted(async () => {
```
Comment (kubajmarek) [inline src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue:128]: Why async if we're not awaiting anything?

File: src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue (hunk 1 above)
Comment (pspaczek) [reply src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue:11]: In this case, we want to use it.

File: src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue (hunk 2 above)
Comment (pspaczek) [reply src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue:128]: typo; thanks

File: src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue (hunk 1 above)
Comment (pspaczek) [reply src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue:11]: this div is our helper; its wrap content to use `<transition>` always

File: src/components/atoms/UiButton/UiButton.vue
Comment (pspaczek) [reply src/components/atoms/UiButton/UiButton.vue:1]: IMO users have a full control on disabling, we should keep this

**Comment by pspaczek** [comment]
> I'm wondering about the behaviour of disabled loading button - I'm not sure whether the loader should be greyed out. It's disabled, as a result of loading, so loading status is still "active". Recent example of non-grayed button [here](https://www.figma.com/file/wz68QdkR6ldjmVrPYrazsr/Core-Triage-UI-Library?node-id=15584%3A126884&mode=dev).

`--is-disabled` just handle the case when the user use `--is-disabled` and set `is-loading`; If you want you can skip it ;) and never use it. But it should be styled IMO

**Comment by kubajmarek** [comment]
> > I'm wondering about the behaviour of disabled loading button - I'm not sure whether the loader should be greyed out. It's disabled, as a result of loading, so loading status is still "active". Recent example of non-grayed button [here](https://www.figma.com/file/wz68QdkR6ldjmVrPYrazsr/Core-Triage-UI-Library?node-id=15584%3A126884&mode=dev).
> 
> `--is-disabled` just handle the case when the user use `--is-disabled` and set `is-loading`; If you want you can skip it ;) and never use it. But it should be styled IMO

The issue is that you'd still want to use `--is-disabled` to remove changes on hover/active. So either way, you'd have to either:
- overwrite background variables
- separately overwrite `--loader-ellipsis-dot-background`

We should probably ask Piotr what should be the look for disabled & loading.

**Comment by kubajmarek** [comment]
> We should probably ask Piotr what should be the look for disabled & loading.

Alternatively, maybe `isLoading` should also add `--is-loading` class modifier, that would work similarly (disable modified hover/active), also change `cursor: wait;`, but not change color scheme.

**Comment by pspaczek** [comment]
> 



> > > I'm wondering about the behaviour of disabled loading button - I'm not sure whether the loader should be greyed out. It's disabled, as a result of loading, so loading status is still "active". Recent example of non-grayed button [here](https://www.figma.com/file/wz68QdkR6ldjmVrPYrazsr/Core-Triage-UI-Library?node-id=15584%3A126884&mode=dev).
> > 
> > 
> > `--is-disabled` just handle the case when the user use `--is-disabled` and set `is-loading`; If you want you can skip it ;) and never use it. But it should be styled IMO
> 
> The issue is that you'd still want to use `--is-disabled` to remove changes on hover/active. So either way, you'd have to either:
> 
> * overwrite background variables
> * separately overwrite `--loader-ellipsis-dot-background`
> 
> We should probably ask Piotr what should be the look for disabled & loading.

when --is-disabled `hover` and `active` don't change anything, has the same color like initial state and this is fine form me. IMO it's not important for us now because we don;t want to use `--is-disabled` with loader nowe. I add it just to avoid colors mismatching

File: src/components/atoms/UiButton/UiButton.vue [hunk 3]
```diff
@@ -4,20 +4,29 @@
     v-keyboard-focus
     v-bind="routeAttrs"
     class="ui-button"
+    :tabindex="isLoading ? -1 : null"
```
Comment (kubajmarek) [inline src/components/atoms/UiButton/UiButton.vue:7]: `isLoading` (and also maybe `--is-disabled` should become `isDisabled`) should make button `aria-disabled`.

File: src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue (hunk 1 above)
Comment (pspaczek) [reply src/components/molecules/UiLoader/_internal/UiLoaderTransition.vue:11]: 👍

**Review by kubajmarek — APPROVED** [review]

**Pull request merged**